src/Microsoft.Net.Http.Headers/NameValueHeaderValue.cs

```diff
@@ -142,6 +143,30 @@ public override bool Equals(object obj)
             }
         }
 
+        public StringSegment GetUnescapedValue()
+        {
+            if (!HeaderUtilities.IsQuoted(_value))
+            {
+                return _value;
+            }
+            return UnescapeAsQuotedString(_value);
+        }
+
+        public void SetAndEscapeValue(StringSegment value)
+        {
+            HeaderUtilities.ThrowIfReadOnly(IsReadOnly);
+            if (StringSegment.IsNullOrEmpty(value) || (GetValueLength(value, 0) == value.Length))
+            {
+                _value = value;
```

Comment:
Setting the Value will throw if there is an error. If the format is invalid, we still need to try calling EscapeAsQuotedString before throwing.